test/e2e/storage/non_graceful_node_shutdown.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2enode "k8s.io/kubernetes/test/e2e/framework/node"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2epv "k8s.io/kubernetes/test/e2e/framework/pv"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	storageframework "k8s.io/kubernetes/test/e2e/storage/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+const (
+	podWaitTimeout = time.Minute * 5
+)
+
+/*
+This test assumes the following:
+- The infra is GCP.
+- NodeOutOfServiceVolumeDetach feature is enabled.
+
+This test performs the following:
+- Deploys a gce-pd csi driver
+- Creates a gce-pd csi storage class
+- Creates a pvc using the created gce-pd storage class
+- Creates an app deployment with replica count 1 and uses the created pvc for volume
+- Shutdowns the kubelet of node on which the app pod is scheduled.
+  This shutdown is a non graceful shutdown as by default the grace period is 0 on Kubelet.
+- Adds `out-of-service` taint on the node which is shut down.
+- Verifies that pod gets immediately scheduled to a different node and gets into running and ready state.
```

Comment:
Can we also verify that when the node comes back up,
1.the kubelet state gets reconciled properly;
2.after the `out-of-service` taint gets removed, the node is ready for workloads as usual?

Reply:
Good point. Thank you

Reply:
Done. [ See line no 131 and 132 in [test/e2e/storage/non_graceful_node_shutdown.go]